lsp/nls/src/background.rs

```diff
@@ -0,0 +1,396 @@
+                            .into_iter()
+                            .flat_map(|e| world.lsp_diagnostics(file_id, e)),
+                    );
+                }
+
+                // If there's been an update to the file, don't send back a stale response.
+                cmds.extend(cmd_rx.try_recv());
+                if !cmds.iter().any(|cmd| match cmd {
+                    Command::UpdateFile { uri, .. } => {
+                        uri_to_path(uri).map_or(false, |p| p == path)
+                    }
+                    _ => false,
+                }) {
+                    response_tx
+                        .send(Response::Diagnostics(Diagnostics { path, diagnostics }))
+                        .ok()?;
+                }
+            }
+        }
+
+        drain_ready(&cmd_rx, &mut cmds);
+        if cmds.is_empty() && evals.is_empty() {
+            // Wait for a command to be available.
+            cmds.push(cmd_rx.recv().ok()?);
+        }
+    }
+}
+
+struct SupervisorState {
+    cmd_rx: Receiver<Command>,
+    response_tx: Sender<Diagnostics>,
+    child: Child,
+    cmd_tx: IpcSender<Command>,
+    response_rx: Receiver<Response>,
+
+    contents: HashMap<Url, String>,
+
+    // If evaluating a file causes the worker to time out or crash, we blacklist that file
+    // and refuse to evaluate it anymore. This could be relaxed (e.g. maybe we're willing to
+    // try again after a certain amount of time?).
+    banned_files: HashSet<Url>,
+    eval_in_progress: Option<(Url, Instant)>,
+}
+
+enum SupervisorError {
+    MainExited,
+    WorkerExited,
+    WorkerTimedOut(Url),
+}
+
+impl SupervisorState {
+    fn spawn_worker() -> anyhow::Result<(Child, IpcSender<Command>, Receiver<Response>)> {
+        let path = std::env::current_exe()?;
+        let (oneshot_server, server_name) =
+            IpcOneShotServer::<(IpcSender<Command>, IpcReceiver<Response>)>::new()?;
+        let child = std::process::Command::new(path)
+            .args(["--main-server", &server_name])
+            .spawn()?;
+
+        let (_, (ipc_cmd_tx, ipc_response_rx)) = oneshot_server.accept()?;
+        let ipc_response_rx = ipc_channel::router::ROUTER
+            .route_ipc_receiver_to_new_crossbeam_receiver(ipc_response_rx);
+        Ok((child, ipc_cmd_tx, ipc_response_rx))
+    }
+
+    fn new(cmd_rx: Receiver<Command>, response_tx: Sender<Diagnostics>) -> anyhow::Result<Self> {
+        let (child, cmd_tx, response_rx) = SupervisorState::spawn_worker()?;
+        Ok(Self {
+            cmd_rx,
+            response_tx,
+            contents: HashMap::new(),
+            child,
+            cmd_tx,
+            response_rx,
+            banned_files: HashSet::new(),
+            eval_in_progress: None,
+        })
+    }
+
+    fn handle_command(&mut self, msg: Command) -> Result<(), SupervisorError> {
+        if let Command::UpdateFile { uri, text } = &msg {
+            self.contents.insert(uri.clone(), text.clone());
```

Comment:
Should we also remove the file from the banned ones on update? Or maybe a timeout is indeed a better way, or a mix of the two (only remove after timeout + update). But that can wait for another PR, anyway.

Reply:
I'm mainly worried about working on files that take a long time to evaluate. The editor might update the contents several times per second, and so if we keep on trying to evaluate and timing out then we're just pegging their CPU for nothing.

I agree that banning the file *forever* is probably too aggressive and I haven't thought too carefully about better policies. I think in the longer term, incremental evaluation would solve this problem: even if evaluation is slow, it's making progress building the incremental cache and so future evals will be fast.